test/e2e/storage/pvc_storageclass.go

```diff
@@ -18,172 +18,172 @@ package storage
-		t = testsuites.StorageClassTest{
-			Timeouts:  f.Timeouts,
-			ClaimSize: "1Gi",
-		}
 	})
 
-	f.Describe("Retroactive StorageClass assignment", framework.WithSerial(), framework.WithDisruptive(), func() {
-		ginkgo.It("should assign default SC to PVCs that have no SC set", func(ctx context.Context) {
-
-			// Temporarily set all default storage classes as non-default
-			restoreClasses := temporarilyUnsetDefaultClasses(ctx, client)
-			defer restoreClasses()
-
-			// Create PVC with nil SC
-			pvcObj := e2epv.MakePersistentVolumeClaim(e2epv.PersistentVolumeClaimConfig{
-				NamePrefix: prefixPVC,
-				ClaimSize:  t.ClaimSize,
-				VolumeMode: &t.VolumeMode,
-			}, namespace)
-			pvc, err = client.CoreV1().PersistentVolumeClaims(pvcObj.Namespace).Create(ctx, pvcObj, metav1.CreateOptions{})
-			framework.ExpectNoError(err, "Error creating PVC")
-			defer func(pvc *v1.PersistentVolumeClaim) {
-				// Remove test PVC
-				err := client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
-				framework.ExpectNoError(err, "Error cleaning up PVC")
-			}(pvc)
-
-			// Create custom default SC
-			storageClass := testsuites.SetupStorageClass(ctx, client, makeStorageClass(prefixSC))
-
-			// Wait for PVC to get updated with the new default SC
-			pvc, err = waitForPVCStorageClass(ctx, client, namespace, pvc.Name, storageClass.Name, f.Timeouts.ClaimBound)
-			framework.ExpectNoError(err, "Error updating PVC with the correct storage class")
-
-			// Create PV with specific class
-			pv := e2epv.MakePersistentVolume(e2epv.PersistentVolumeConfig{
-				NamePrefix:       "pv-",
-				StorageClassName: storageClass.Name,
-				VolumeMode:       pvc.Spec.VolumeMode,
-				PVSource: v1.PersistentVolumeSource{
-					HostPath: &v1.HostPathVolumeSource{
-						Path: "/tmp/test",
-					},
-				},
-			})
-			_, err = e2epv.CreatePV(ctx, client, f.Timeouts, pv)
-			framework.ExpectNoError(err, "Error creating pv %v", err)
-			ginkgo.DeferCleanup(e2epv.DeletePersistentVolume, client, pv.Name)
-
-			// Verify the PVC is bound and has the new default SC
-			claimNames := []string{pvc.Name}
-			err = e2epv.WaitForPersistentVolumeClaimsPhase(ctx, v1.ClaimBound, client, namespace, claimNames, 2*time.Second /* Poll */, t.Timeouts.ClaimProvisionShort, false)
-			framework.ExpectNoError(err)
-			updatedPVC, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvc.Name, metav1.GetOptions{})
-			framework.ExpectNoError(err)
-			gomega.Expect(*updatedPVC.Spec.StorageClassName).To(gomega.Equal(storageClass.Name), "Expected PVC %v to have StorageClass %v, but it has StorageClass %v instead", updatedPVC.Name, prefixSC, updatedPVC.Spec.StorageClassName)
-			framework.Logf("Success - PersistentVolumeClaim %s got updated retroactively with StorageClass %v", updatedPVC.Name, storageClass.Name)
-		})
+	f.It("should assign default StorageClass to PVCs retroactively", f.WithDisruptive(), f.WithSerial(), func(ctx context.Context) {
```

Comment:
How exactly does this change help with **suspected** race with other tests? It still relies on modifying existing storage classes for the duration of the test, and also reverts SC settings via `defer`. The only thing changing is that here we don't check for specific storage class name being assigned to PVC making the test even less predictable as it would tolerate any SC being assigned - which should never happen during execution of this test. Seems like we're masking the problem without knowing the root cause.

If this test is serial and disruptive, how is it possible that it seems to be racy? Did we rule out the Prow config?

Reply:
Thank you for the review Roman! The latest revision now validates the StorageClass assignment as well.

> If this test is serial and disruptive, how is it possible that it seems to be racy? Did we rule out the Prow config?

The following logs can be observed in every failure:
```
0820 06:06:25.981634 15020 pvc_storageclass.go:178] PersistentVolumeClaim retro-pvc-lmzlm has unexpected StorageClass balanced-csi, expected StorageClass is: retromr8np
```

`balanced-csi` StorageClass is created by kOps during bootstrap. I believe we should be able to disable this via the `ManageStorageClasses` [parameter](https://github.com/kubernetes/kops/blob/e7be8333165b5494b00a9de634c5d687ae67daab/pkg/apis/kops/componentconfig.go#L931). Doing this, there should be no need to unset default storage classes in the test itself. 

We can either go with that approach, or introduce additional validation in the test here to address this race. As I understand the issue, it is possible for the [StorageClass lister ](https://github.com/kubernetes/kubernetes/blob/b2031b3cb46e946ee72eab7bda87b046db138d62/plugin/pkg/admission/storage/storageclass/setdefault/admission.go#L108)to return cached data if the PVC gets created immediately after the SC label has been updated.

Would love to hear your thoughts on this.

Reply:
Those are great points! I don't know much about kOps but was suspecting it of reconciling the `balanced-csi` storage class - but so far I could not find it and it seems like it's a one off action - and if it is then introducing additional validation would reveal it (test should timeout and fail if it can't ensure no SCs are default). I think this should be done anyway as it makes the test more robust.

Disabling the SC with `ManageStorageClasses` seems like a pretty good idea and I think you're right, in that case we should get a cluster with no default SCs, but relying on that in the test does not seem right. If somehow it would get re-enabled in the future, the failures would be reintroduced. Also, this might break other test, maybe some of them already rely on having a default SC :shrug:

So I'd first try the additional validation, and if the test starts failing due to SC being periodically reconciled go with disabling `ManageStorageClasses`. WDYT?

Reply:
Awesome, that sounds like a good plan to me, agree with all of the above. I'll put up a new revision later today and cc you when this PR is ready for review again.